### PR TITLE
[extension-selenium] Update screenshot taking step (#1795)

### LINKED ISSUE #1795: Update screenshot taking step to publish screenshot only in report
`When I take screenshot` - should publish the screenshot to the report only
``` When I take screenshot and save it to folder `$path` ``` -  should publish screenshot only to the folder

### PR DESCRIPTION
**Pull request by draker94**
Closes #1795.
 
- Update screenshot taking step (`When I take screenshot`) to publish screenshots in the report only.

### CODE REVIEW

File: docs/modules/plugins/partials/generic-ui-steps.adoc [hunk 1]
```diff
@@ -297,11 +297,11 @@ Then elements located `By.xpath(//span[@id='item-to-purchase'])` are sorted by t
 ----
 
 
-=== Take a screenshot
+=== Take a screenshot in Allure Report
```
Comment (valfirst) [inline docs/modules/plugins/partials/generic-ui-steps.adoc:300]:
```suggestion
=== Take a screenshot
```
we may use different reporting engines, so UI steps should not declare any explicit reporting engine

File: docs/modules/plugins/partials/generic-ui-steps.adoc [hunk 2]
```diff
@@ -297,11 +297,11 @@ Then elements located `By.xpath(//span[@id='item-to-purchase'])` are sorted by t
...
 
-Takes a screenshot and saves it to the default folder defined by the property `ui.screenshots.directory`.
+Takes a screenshot and publish it to the allure report.
```
Comment (valfirst) [inline docs/modules/plugins/partials/generic-ui-steps.adoc:302]:
```suggestion
Takes a screenshot and publish it to the report.
```

File: docs/modules/plugins/partials/generic-ui-steps.adoc [hunk 3]
```diff
@@ -297,11 +297,11 @@ Then elements located `By.xpath(//span[@id='item-to-purchase'])` are sorted by t
...
 
-TIP: The full path to the screenshot is posted to the logs and the report.
+TIP: In the report screenshot will be posted with additional information like date and browser.
```
Comment (valfirst) [inline docs/modules/plugins/partials/generic-ui-steps.adoc:304]:
```suggestion
```
these are implementation details, no need to put it to the docs

File: vividus-extension-selenium/src/main/java/org/vividus/selenium/screenshot/AbstractScreenshotTaker.java [hunk 4]
```diff
@@ -94,14 +97,23 @@ public Path takeScreenshot(Path screenshotFilePath) throws IOException
...
         return null;
     }
 
+    @Override
+    public void takeScreenshotInAllure(String screenshotName)
+    {
+        String screenshotPublishedName = generateScreenshotFileName(screenshotName);
+        Attachment attachment = new Attachment(takeScreenshotAsByteArray(), screenshotPublishedName);
+        eventBus.post(new AttachmentPublishEvent(attachment));
+        LOGGER.atInfo().addArgument(screenshotPublishedName).log(SCREENSHOT_TAKEN_LOG_MESSAGE);
```
Comment (valfirst) [inline vividus-extension-selenium/src/main/java/org/vividus/selenium/screenshot/AbstractScreenshotTaker.java:114]:
```suggestion
```
redundant, this info will be in report any way

File: vividus-extension-selenium/src/main/java/org/vividus/selenium/screenshot/ScreenshotTaker.java [hunk 5]
```diff
@@ -30,4 +30,6 @@
     Path takeScreenshotAsFile(String screenshotName) throws IOException;
 
     Path takeScreenshot(Path screenshotFilePath) throws IOException;
+
+    void takeScreenshotInAllure(String screenshotName);
```
Comment (valfirst) [inline vividus-extension-selenium/src/main/java/org/vividus/selenium/screenshot/ScreenshotTaker.java:34]: why is new method needed? why can't we use `Optional<Screenshot> takeScreenshot(String screenshotName);`?

File: vividus-extension-selenium/src/main/java/org/vividus/selenium/screenshot/AbstractScreenshotTaker.java [hunk 6]
```diff
@@ -94,14 +97,23 @@ public Path takeScreenshot(Path screenshotFilePath) throws IOException
...
             return screenshotFilePath;
         }
         return null;
     }
 
+    @Override
+    public void takeScreenshotInAllure(String screenshotName)
+    {
+        String screenshotPublishedName = generateScreenshotFileName(screenshotName);
+        Attachment attachment = new Attachment(takeScreenshotAsByteArray(), screenshotPublishedName);
```
Comment (valfirst) [inline vividus-extension-selenium/src/main/java/org/vividus/selenium/screenshot/AbstractScreenshotTaker.java:112]: it makes sense to provide content type as well

**Review by valfirst — CHANGES_REQUESTED** [review]

File: vividus-extension-selenium/src/main/java/org/vividus/steps/ui/ScreenshotTakingSteps.java [hunk 7]
```diff
@@ -19,23 +19,35 @@
...
+    }
 
     /**
-     * Takes a screenshot and saves it to the default location
-     * @throws IOException If an input or output exception occurred
+     * Takes a screenshot and publish it to the report
      */
     @When("I take screenshot")
-    public void whenITakeScreenshot() throws IOException
+    public void whenITakeScreenshot()
```
Comment (valfirst) [inline vividus-extension-selenium/src/main/java/org/vividus/steps/ui/ScreenshotTakingSteps.java:44]:
```suggestion
    public void takeScreenshot()
```
I know it's a bit out of scope, but it's a good chance to change this ugly name

File: vividus-extension-selenium/src/main/java/org/vividus/selenium/screenshot/AbstractScreenshotTaker.java [hunk 8]
```diff
@@ -97,6 +87,15 @@ public Path takeScreenshot(Path screenshotFilePath) throws IOException
...
     }
 
+    @Override
+    public Optional<Screenshot> takeScreenshot(String screenshotName)
+    {
+        byte[] screenshotData = takeScreenshotAsByteArray();
+        return screenshotData.length > 0
+                ? Optional.of(new Screenshot(generateScreenshotFileName(screenshotName), screenshotData))
+                : Optional.empty();
+    }
```
Comment (valfirst) [inline vividus-extension-selenium/src/main/java/org/vividus/selenium/screenshot/AbstractScreenshotTaker.java:97]: what's the reason to add a default implementation? Web and Mobile App screenshot takers have this method implemented, but in different ways

File: vividus-extension-selenium/src/main/resources/spring.xml [hunk 9]
```diff
@@ -138,13 +138,7 @@
 
     <bean class="org.vividus.ui.action.search.IndexElementFilter" />
 
-    <bean class="org.vividus.selenium.screenshot.AbstractScreenshotTaker" id="abstractScreenshotTaker" abstract="true">
-        <property name="screenshotDirectory">
-            <bean class="java.io.File">
-                <constructor-arg value="${ui.screenshot.directory}" />
```
Comment (valfirst) [inline vividus-extension-selenium/src/main/resources/spring.xml:144]: it makes sense to remove the default property as well: https://github.com/vividus-framework/vividus/blob/b329f3c1ffdd5d71e40f3f70800f76d08a789d6e/vividus-extension-selenium/src/main/resources/properties/defaults/default.properties#L25

**Review by valfirst — CHANGES_REQUESTED** [review]

**Review by valfirst — APPROVED** [review]

**Review by ikalinin1 — APPROVED** [review]

**Review by uarlouski — APPROVED** [review]

**Pull request merged**